### PR TITLE
fix(optimizer): fix cardinality visitor -- filter

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- The `logical_pk` is empty() means there is no pk.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/frontend/src/optimizer/plan_visitor/cardinality_visitor.rs [hunk 1]
```diff
@@ -55,7 +55,7 @@ impl CardinalityVisitor {
 
         if unique_keys
             .iter()
-            .any(|unique_key| eq_set.is_superset(unique_key))
+            .any(|unique_key| eq_set.is_superset(unique_key) && !unique_key.is_empty())
```
Comment (BugenZhao) [inline src/frontend/src/optimizer/plan_visitor/cardinality_visitor.rs:58]: What about using `filter` here?

**Review by BugenZhao — COMMENTED** [review]
This is interesting. Theoretically, a relation without a primary key should have a cardinality of 0..=1. However, it seems we also leverage "empty primary key" in the optimizer to express that the primary key is lost or unknown (the output of this `Project` in the test case), so we have this issue. 🤔

Is there any way to distinguish between them?


```
......
     └─LogicalFilter { predicate: (CorrelatedInputRef { index: 0, correlated_id: 1 } = 1:Int32) }
       └─LogicalProject { exprs: [1:Int32] }
         └─LogicalScan { table: t2, columns: [j, _row_id] }
```

**Comment by chenzl25** [comment]
> Theoretically, a relation without a primary key should have a cardinality of 0..=1

This is not quite accurate. A relation without a primary key could have any number of rows. 

> we also leverage "empty primary key" in the optimizer to express that the primary key is lost or unknown

Actually, I haven't seen a relation that could contain an empty primary key before. If the relation could guarantee its cardinality is 0..=1 and then I think it can use any one of its column as a primary key.

**Comment by BugenZhao** [comment]
> This is not quite accurate. A relation without a primary key could have any number of rows.

What I was trying to say is "a relation with an empty primary key". This seems to be like using a tuple without any elements, or unit type `()` as the key of a map. Then it's sure that the map will have at most one item. Currently, we don't distinguish between an empty key and no key. 🤔

> Actually, I haven't seen a relation that could contain an empty primary key before.

IIUC, the output of simple aggregation has an empty primary key?

**Comment by chenzl25** [comment]
> > This is not quite accurate. A relation without a primary key could have any number of rows.
> 
> What I was trying to say is "a relation with an empty primary key". This seems to be like using a tuple without any elements, or unit type `()` as the key of a map. Then it's sure that the map will have at most one item. Currently, we don't distinguish between an empty key and no key. 🤔
> 
> > Actually, I haven't seen a relation that could contain an empty primary key before.
> 
> IIUC, the output of simple aggregation has an empty primary key?

Make senses. So pk should be an `Option<Vec<usize>>` instead?

**Pull request merged**